### PR TITLE
makefiles/utils/{strings, paths}: Add function library.

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

Add a set of functions written in pure makefile. The only function that is really useful now is relpath, to calculate relative paths in systems without coreutils/realpath (e.g. Mac).

Tests were added to tests/build_system_utils. The Makefile for the test was also reworked to avoid code repetition.

### Testing procedure

Run `make -C tests/build_system_utils`. You should see several green lines showing tests that passed.

### Issues/PRs references

~Needed for #10195 .~

### CODE REVIEW

File: makefiles/utils/strings.mk [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# Return a string ("x") if the argument is empty and vice versa
+strnot = $(if $(1),,x)
+
+# Concatenate a list of strings
+# Defined via "?=" because redefinition kill the recursive definition
+concat ?= $(if $(1),$(firstword $(1))$(call concat,$(call rest,$(1))),)
+
+# Intercalate: concatenate words in a list with a separator between each one.
+intercal = $(firstword $(2))$(call concat,$(addprefix $(1),$(call rest,$(2))))
```
Comment (cladmi) [inline makefiles/utils/strings.mk]: By taking the list as first argument and the separator as second, concat could be the same "public" function I think.

```
Concatenate a list of strings:

$1: list of strings
$2: join str, default to empty
```

Maybe for the implementation you still need the other one but only one public would be enough.

File: makefiles/utils/test-checks.mk [hunk 2]
```diff
@@ -5,3 +5,21 @@ test-ensure_value:
...
+	@test -z "$(call assert,$@,"This should not fail")"
+
+test-assert-negative:
+	@test -z "$(call assert,$^,"This should fail")"
+
+test-assert_not:
+	@test -z "$(call assert_not,$^,"This should not fail")"
+
+test-assert_not-negative:
+	@test -z "$(call assert_not,$@,"This should fail")"
```
Comment (cladmi) [inline makefiles/utils/test-checks.mk]: Not working, when calling it with `assert_not,,` the test succeeds anyway.

File: tests/build_system_utils/Makefile [hunk 3]
```diff
@@ -5,29 +5,77 @@ include $(RIOTBASE)/Makefile.include
 
 
 # Test utils commands
+
+ok_message = $(COLOR_ECHO) "$(COLOR_GREEN) [OK] $1 $(COLOR_RESET)"
+fail_message = $(COLOR_ECHO) "$(COLOR_RED) [FAIL] $1 $(COLOR_RESET)"
+
 define command_should_fail
-$1 2>/dev/null && { echo "Command '$1' should have failed but did not" >&2; $1; exit 1; } || true
+$1 2>/dev/null && { $(call fail_message,Command '$1' should have failed but did not) >&2; $1; exit 1; } || $(call ok_message,$2)
```
Comment (cladmi) [inline tests/build_system_utils/Makefile]: This does not work, none of the `negative` functions detect when something does not fail.

Comment (cladmi) [reply tests/build_system_utils/Makefile]: With this the test still pass:

``` diff
diff --git a/makefiles/utils/test-checks.mk b/makefiles/utils/test-checks.mk
index 3ee960cfd..f99a3d12e 100644
--- a/makefiles/utils/test-checks.mk
+++ b/makefiles/utils/test-checks.mk
@@ -4,7 +4,7 @@ test-ensure_value:
        @test "$@" = "$(call ensure_value,$@,"This should not fail")"

 test-ensure_value-negative:
-       @echo $(call ensure_value,$^,"This should fail")
+       true

 test-assert:
        @test -z "$(call assert,$@,"This should not fail")"
```

**Comment by cladmi** [comment]
Oops, my bad, I used "add comment" instead of "add to review" and even edited the old ones as thought it was in a review… :disappointed:

File: makefiles/utils/strings.mk [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# String handling functions
+
+# Return an empty string if the arguments are different (works only for single
+# words
+streq = $(filter $(1),$(2))
```
Comment (cladmi) [inline makefiles/utils/strings.mk]: Even with only single words, it does not work. Using `findstring` would be better to not match patterns.
Here it could be checked that indeed `words $(1) == 1` and `words $(2) == 1`.

File: makefiles/utils/strings.mk [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+# Get the remaining words in a list (i.e. everything except the first)
+rest = $(wordlist 2,$(words $(1)),$(1))
+
+# Strings as truth values
+# In make, empty strings function as a "false" value and non-empty ones
+# as true.
+
+# Return a string ("x") if the argument is empty and vice versa
+strnot = $(if $(1),,x)
```
Comment (cladmi) [inline makefiles/utils/strings.mk]: The `if` is somehow clear enough and `strnot` is unused.

**Comment by jcarrano** [comment]
@cladmi  I stated to fix some of the stuff, but each time I approach this I cannot help feeling I'm reinventing the wheel, or in this case, recursive data types, and if I were to keep on going I'd probably end up defining folds, and zips, and maps, and a whole language on top of Make.

On the other hand, since we _don't_ want a configure step (and we have our good reasons) we need to have quite a bit of functionality in make.

Now, looking around I see that Make with Guile (i.e. scheme) support [**is available in ubuntu**](https://packages.ubuntu.com/xenial/make-guile). Writing all this complex logic in scheme is much easier. Even for someone who does not know programming (read [SICP](https://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book.html)).

There's still the issue of Mac users (Windows users can run Ubuntu via WLS). AFAIK the newer versions of Make **are available on mac** ([via homebrew](https://formulae.brew.sh/formula/make#default)). 

In conclusion:

- We should implement these functions using scheme.
- Switch the ubuntu image from make to make-guile.
- Mac users: if you want the build to run natively in your platform, please make a small effort and setup your system with up-to-date tools.

**Comment by cladmi** [comment]
Just keep the functions we currently need that are hard to do directly in make.

Some of the utils are not explicitely needed `strnot` and `streq` are just replacing what is normally done in `make` with a different name.

If the new needed one is currently only `relpath` for thin-archives we can just first use the binary/a python version and update if there is really a performance issue.
